src/p11_child/p11_child_openssl.c

```diff
@@ -0,0 +1,1133 @@
+
+    *p11_ctx = ctx;
+
+    return EOK;
+}
+
+static int talloc_free_x509_store(struct p11_ctx *p11_ctx)
+{
+    X509_STORE_free(p11_ctx->x509_store);
+
+    return 0;
+}
+
+errno_t init_verification(struct p11_ctx *p11_ctx,
+                          struct cert_verify_opts *cert_verify_opts)
+{
+    int ret;
+    X509_STORE *store = NULL;
+    unsigned long err;
+    X509_LOOKUP *lookup = NULL;
+
+    store = X509_STORE_new();
+    if (store == NULL) {
+        err = ERR_get_error();
+        DEBUG(SSSDBG_OP_FAILURE, "X509_STORE_new failed [%lu][%s].\n",
+                                 err, ERR_error_string(err, NULL));
+        return ENOMEM;
+    }
+
+    lookup = X509_STORE_add_lookup(store, X509_LOOKUP_file());
+    if (lookup == NULL) {
+        err = ERR_get_error();
+        DEBUG(SSSDBG_OP_FAILURE, "X509_LOOKUP_file failed [%lu][%s].\n",
+                                 err, ERR_error_string(err, NULL));
+        ret = EIO;
+        goto done;
+    }
+
+    if (!X509_LOOKUP_load_file(lookup, p11_ctx->ca_db, X509_FILETYPE_PEM)) {
+        err = ERR_get_error();
+        DEBUG(SSSDBG_OP_FAILURE, "X509_LOOKUP_load_file failed [%lu][%s].\n",
+                                 err, ERR_error_string(err, NULL));
+        ret = EIO;
+        goto done;
+    }
+
+    p11_ctx->x509_store = store;
+    talloc_set_destructor(p11_ctx, talloc_free_x509_store);
+
+    ret = EOK;
+
+done:
+    if (ret != EOK) {
+        X509_STORE_free(store);
+        X509_LOOKUP_free(lookup);
+    }
+
+    return ret;
+}
+
+static int b64_to_cert(const char *b64, X509 **cert)
+{
+    X509 *x509;
+    unsigned char *der = NULL;
+    const unsigned char *d;
+    size_t der_size;
+
+    der = sss_base64_decode(NULL, b64, &der_size);
+    if (der == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "sss_base64_decode failed.\n");
+        return ENOMEM;
+    }
+
+    d = (const unsigned char *) der;
+    x509 = d2i_X509(NULL, &d, (int) der_size);
+    talloc_free(der);
+    if (x509 == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "d2i_X509 failed.\n");
+        return EINVAL;
+    }
+
+    *cert = x509;
+
+    return 0;
+}
+
+bool do_verification(struct p11_ctx *p11_ctx, X509 *cert)
+{
+    bool res = false;
+    int ret;
+    X509_STORE_CTX *ctx = NULL;
+    unsigned long err;
+
+    ctx = X509_STORE_CTX_new();
+    if (ctx == NULL) {
+        err = ERR_get_error();
+        DEBUG(SSSDBG_OP_FAILURE, "X509_STORE_CTX_new failed [%lu][%s].\n",
+                                 err, ERR_error_string(err, NULL));
+        return false;
+    }
+
+    if (!X509_STORE_CTX_init(ctx, p11_ctx->x509_store, cert, NULL)) {
+        err = ERR_get_error();
+        DEBUG(SSSDBG_OP_FAILURE, "X509_STORE_CTX_init failed [%lu][%s].\n",
+                                 err, ERR_error_string(err, NULL));
+        goto done;
+    }
+
+    ret = X509_verify_cert(ctx);
+    if (ret != 1) {
+        DEBUG(SSSDBG_OP_FAILURE, "X509_verify_cert failed [%d].\n", ret);
```

Comment:
Typically ret would be 0 in case of an error but the man page says 'in exceptional circumstances it can also return a negative code'. So there might be a small chance that the code might have additional information in the case of an error.